src/java.base/share/classes/java/lang/reflect/code/bytecode/BytecodeLift.java

```diff
@@ -418,73 +122,332 @@ public static CoreOps.FuncOp liftToBytecodeDialect(byte[] classdata, String meth
+                            case StackMapFrameInfo.SimpleVerificationTypeInfo.ITEM_DOUBLE ->
+                                b.parameter(TypeDesc.DOUBLE);
+                            case StackMapFrameInfo.SimpleVerificationTypeInfo.ITEM_LONG ->
+                                b.parameter(TypeDesc.LONG);
+                            case StackMapFrameInfo.SimpleVerificationTypeInfo.ITEM_NULL -> // @@@
+                                b.parameter(TypeDesc.J_L_OBJECT);
+                            case StackMapFrameInfo.SimpleVerificationTypeInfo.ITEM_UNINITIALIZED_THIS ->
+                                b.parameter(TypeDesc.ofNominalDescriptor(bytecodeBody.methodModel.parent().get().thisClass().asSymbol()));
+                            case StackMapFrameInfo.ObjectVerificationTypeInfo i ->
+                                b.parameter(TypeDesc.ofNominalDescriptor(i.classSymbol()));
+                            case StackMapFrameInfo.UninitializedVerificationTypeInfo i -> {
+                                // @@@
+                                // label designates the NEW instruction that created the uninitialized value
+                            }
+                            case StackMapFrameInfo.SimpleVerificationTypeInfo.ITEM_TOP ->
+                                throw new IllegalArgumentException("Unexpected item TOP at frame stack.");
                         }
                     }
-                    b.op(frame);
                 }
+                count++;
+                c.blockMap.put(basicBlock, b);
+            }
+
+            // @@@ Needs to be cloned when there are two or more successors
+            Map<Integer, Op.Result> locals = new HashMap<>();
 
-                liftBytecodeBlock(bcr, bcb, b, c);
+            // Map Block arguments to local variables
+            int lvm = 0;
+            for (Block.Parameter bp : entryBlock.parameters()) {
+                // @@@ Reference type
+                Op.Result local = entryBlock.op(CoreOps.var(Integer.toString(lvm), bp));
+                locals.put(lvm++, local);
             }
-        });
 
-        return f;
-    }
+            // Process blocks
+            for (BytecodeBasicBlock bcb : bytecodeBody.blocks) {
+                Block.Builder b = c.blockMap.get(bcb);
+                // If a non-entry block has parameters then they correspond to stack arguments
+                // Push back block parameters on the stack, in reverse order to preserve order on the stack
+                Deque<Value> stack = new ArrayDeque<>();
+                if (!b.parameters().isEmpty() && !b.isEntryBlock()) {
+                    for (int i = b.parameters().size() - 1; i >= 0; i--) {
+                        stack.push(b.parameters().get(i));
+                    }
+                }
+                int ni = bcb.instructions.size();
+                for (int i = 0; i < ni; i++) {
+                    switch (bcb.instructions.get(i)) {
```

Comment:
I think it would be beneficial to separate out the switch over an instruction into a separate method, so the loop logic (over blocks and instructions) can be easily read.

In many cases instructions push an operation result onto the stack and we could factor out that code, where we return a non-null result in such cases. This would also make the code easier to read in the switch cases.

Reply:
I'm actually thinking of refactoring the whole process into a single pass.

Theoretically there should be no reason to prepare blocks and feed instructions into an auxiliary data structure in one pass, add block parameters from exception handlers and from stack map frames in another pass and convert instructions into ops in another pass.